src/abstractdatatable/abstractdatatable.jl

```diff
@@ -706,83 +710,55 @@ Base.hcat(dt1::AbstractDataTable, dt2::AbstractDataTable) = hcat!(dt[:, :], dt2)
 
-# vcat only accepts DataTables. Finds union of columns, maintaining order
-# of first dt. Missing data become null values.
+"""
+    vcat(dts::AbstractDataTable...)
 
-Base.vcat(dt::AbstractDataTable) = dt
+Vertically concatenate `AbstractDataTables` that have the same column names in
+the same order.
 
-Base.vcat(dts::AbstractDataTable...) = vcat(AbstractDataTable[dts...])
+```julia
+julia> dt1 = DataTable(A=1:3, B=1:3);
 
-function Base.vcat{T<:AbstractDataTable}(dts::Vector{T})
-    isempty(dts) && return DataTable()
-    coltyps, colnams, similars = _colinfo(dts)
-
-    res = DataTable()
-    Nrow = sum(nrow, dts)
-    for j in 1:length(colnams)
-        colnam = colnams[j]
-        col = similar(similars[j], coltyps[j], Nrow)
-
-        i = 1
-        for dt in dts
-            if haskey(dt, colnam)
-                copy!(col, i, dt[colnam])
-            end
-            i += size(dt, 1)
-        end
+julia> dt2 = DataTable(A=4:6, B=4:6);
 
-        res[colnam] = col
-    end
-    res
-end
+julia> dt3 = DataTable(A=7:9, B=7:9, C=7:9);
 
-_isnullable{T}(::AbstractArray{T}) = T <: Nullable
-const EMPTY_DATA = NullableArray(Void, 0)
-
-function _colinfo{T<:AbstractDataTable}(dts::Vector{T})
-    dt1 = dts[1]
-    colindex = copy(index(dt1))
-    coltyps = eltypes(dt1)
-    similars = collect(columns(dt1))
-    nonnull_ct = Int[_isnullable(c) for c in columns(dt1)]
-
-    for i in 2:length(dts)
-        dt = dts[i]
-        for j in 1:size(dt, 2)
-            col = dt[j]
-            cn, ct = _names(dt)[j], eltype(col)
-            if haskey(colindex, cn)
-                idx = colindex[cn]
-
-                oldtyp = coltyps[idx]
-                if !(ct <: oldtyp)
-                    coltyps[idx] = promote_type(oldtyp, ct)
-                    # Needed on Julia 0.4 since e.g.
-                    # promote_type(Nullable{Int}, Nullable{Float64}) gives Nullable{T},
-                    # which is not a usable type: fall back to Nullable{Any}
-                    if VERSION < v"0.5.0-dev" &&
-                       coltyps[idx] <: Nullable && !isa(coltyps[idx].types[2], DataType)
-                        coltyps[idx] = Nullable{Any}
-                    end
-                end
-                nonnull_ct[idx] += !_isnullable(col)
-            else # new column
-                push!(colindex, cn)
-                push!(coltyps, ct)
-                push!(similars, col)
-                push!(nonnull_ct, !_isnullable(col))
-            end
-        end
-    end
+julia> vcat(dt1, dt2)
+6×2 DataTables.DataTable
+│ Row │ A │ B │
+├─────┼───┼───┤
+│ 1   │ 1 │ 1 │
+│ 2   │ 2 │ 2 │
+│ 3   │ 3 │ 3 │
+│ 4   │ 4 │ 4 │
+│ 5   │ 5 │ 5 │
+│ 6   │ 6 │ 6 │
 
-    for j in 1:length(colindex)
-        if nonnull_ct[j] < length(dts) && !_isnullable(similars[j])
-            similars[j] = EMPTY_DATA
+julia> vcat(dt1, dt2, dt3)
+ERROR: ArgumentError: columns (A, B) of input(s) (1, 2) != columns (A, B, C) of input(s) (3)
```

Comment:
Yea, you asked me to change this before too and I'm still struggling to think of a better way. This will handle all conditions for any number of inputs. Ideally, we would present the differences in a format like a git diff where only differences are shown and (bonus feature:) they would be shown colorized (red for missing, green for extra columns).

```julia
julia> dt1 = DataTable(A = 1, B = 1)
1×2 DataTables.DataTable
│ Row │ A │ B │
├─────┼───┼───┤
│ 1   │ 1 │ 1 │

julia> dt2 = DataTable(B = 1, A = 1)
1×2 DataTables.DataTable
│ Row │ B │ A │
├─────┼───┼───┤
│ 1   │ 1 │ 1 │

julia> dt3 = DataTable(B = 1, A = 1, C = 1)
1×3 DataTables.DataTable
│ Row │ B │ A │ C │
├─────┼───┼───┼───┤
│ 1   │ 1 │ 1 │ 1 │

julia> vcat(dt1, dt2)
ERROR: ArgumentError: columns (A, B) of input(s) (1) != columns (B, A) of input(s) (2)
Stacktrace:
 [1] vcat(::DataTables.DataTable, ::DataTables.DataTable) at /Users/Cameron/.julia/v0.6/DataTables/src/abstractdatatable/abstractdatatable.jl:756

julia> vcat(dt2, dt3)
ERROR: ArgumentError: columns (B, A) of input(s) (1) != columns (B, A, C) of input(s) (2)
Stacktrace:
 [1] vcat(::DataTables.DataTable, ::DataTables.DataTable) at /Users/Cameron/.julia/v0.6/DataTables/src/abstractdatatable/abstractdatatable.jl:756

julia> vcat(dt1, dt2, dt3)
ERROR: ArgumentError: columns (A, B) of input(s) (1) != columns (B, A) of input(s) (2) != columns (B, A, C) of input(s) (3)
Stacktrace:
 [1] vcat(::DataTables.DataTable, ::DataTables.DataTable, ::DataTables.DataTable, ::Vararg{DataTables.DataTable,N} where N) at /Users/Cameron/.julia/v0.6/DataTables/src/abstractdatatable/abstractdatatable.jl:756
```